src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,142 @@
+    // Adds overflow checks per statement that contains an expression and assume statements in the beginning
+    // specifying the bounds of each argument
+    case f@Function(name, args, results, pres, posts, body) =>
+      Function(name, args, results, pres, posts, body map (computeNewBody(args, _)))(f.info)
+
+    // same as functions
+    case m@Method(receiver, name, args, results, pres, posts, body) =>
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPre(args)
+        val newPost = posts ++ Vector(getPureBlockPost(expr, results))
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    // TODO: should we add overflow checks to predicates?
+
+    case x => x
+  }
+
+  /**
+    * Adds overflow checks to the bodies of a block. If block is parametrized,
+    * `computeNewBody` adds assumptions about the parameters' bounds
+    */
+  private def computeNewBody(args: Vector[Parameter.In], body: Block): Block = {
+    val assumeArgBounds = args map { case p@Parameter.In(_, typ) => Assume(assertionExprInBounds(p, typ)(p.info))(p.info) }
+    val blockStmts = body.stmts map stmtTransform
+    Block(body.decls, assumeArgBounds ++ blockStmts)(body.info)
+  }
+
+  /**
+    * Computes the pre-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that each argument is within the bounds for its type
+    */
+  private def getPureBlockPre(args: Vector[Parameter.In]): Vector[Assertion] = {
+    args.map { p: Parameter.In => assertionExprInBounds(p, p.typ)(p.info) }
+  }
+
+  /**
+    * Computes the post-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that the expression result is within the bounds of its type
+    */
+  private def getPureBlockPost(body: Expr, results: Vector[Parameter.Out]): Assertion = {
+    // relies on the current assumption that pure functions and methods must have exactly one result argument
+    if (results.length != 1) violation("Pure functions and methods must have exactly one result argument")
+    assertionExprInBounds(body, results(0).typ)(addAnnotation(body.info))
+  }
+
+  private def stmtTransform(stmt: Stmt): Stmt = stmt match {
```

Comment:
I also thought about this and, in the end, I chose to go with adding overflow checks to statements because it is not always possible to add an overflow check in the scope that an expression appears. For example, when you have a pure function, you cannot add an assertion to its body to check that the result of its expression is within bounds because the body can only contain one expression. As such, I chose to add overflow checks in the scopes that I see fit (such as specific statements (e.g. If statements, while loops, assignments) and also, pure methods and functions (by adding the overflow checks to their contracts) and regular methods and functions. So, if the expression that you provided appears in such a context such as an assignment
```
x := foo(x + 202020020202020)
```
then it will be detected as `x + 202020020202020` is a subexpression of ` foo(x + 202020020202020)`. Otherwise, it wont be detected

Reply:
With respect to what was said yesterday at the meeting, I realized that we don't have an explicit ghost modifier for statements (that I am aware of) and as such, this may lead to unexpected results when we increment a ghost variable, as this will be treated as a regular assignment and overflow checks will be generated. Should we add one?

Reply:
There are ghost stmts in Gobra, you can simply put `ghost` in front of a stmt (see `ghostStatement` in the parser).
For now I would create overflow checks for ghost stmts as well as it seems very related to the discussion of mathematical types

Reply:
But not at the level of the internal language right? I may have missed something

Reply:
No I think you're right. It looks like this information is lost during desugaring

Reply:
I am confused, to my knowledge, we want to generate the bound checks also for ghost code, so there should not be a problem.

Reply:
Yes that was my take away message from the meeting as well

Reply:
Sorry, my bad